### PR TITLE
HOSTEDCP-1611: Add Hypershift job for AKS MC + external DNS HC

### PR DESCRIPTION
**Pull request by fxierh**
Thank you, @jinyunma, for your suggestions. They were very helpful. 

Successful rehearsals:
https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/53175/rehearse-53175-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7/1801704590970720256
https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/53175/rehearse-53175-periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-ovn-hypershift-public-private-mgmt-f7/1801704590882639872
https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/53175/rehearse-53175-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-aws-ipi-ovn-hypershift-guest-f7/1801704590777782272
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/53175/rehearse-53175-periodic-ci-openshift-hypershift-release-4.16-periodics-e2e-azure-ovn-conformance/1801704590912000000
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/53175/rehearse-53175-periodic-ci-openshift-hypershift-release-4.17-periodics-e2e-azure-conformance/1802995092131155968

### CODE REVIEW

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-aws-ipi-ovn-hypershift-guest-f7 periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-ovn-hypershift-public-private-mgmt-f7 periodic-ci-openshift-hypershift-release-4.16-periodics-e2e-azure-conformance

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.16-periodics-e2e-azure-ovn-conformance

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-aws-ipi-ovn-hypershift-guest-f7 periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-nightly-aws-ipi-ovn-hypershift-public-private-mgmt-f7 periodic-ci-openshift-hypershift-release-4.16-periodics-e2e-azure-ovn-conformance periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-guest-f7

**Comment by fxierh** [comment]
/cc @sjenning 
or
/cc @bryan-cox 
or
/cc @Patryk-Stefanski 

Could you PTAL ? Thank you.

**Comment by fxierh** [comment]
/cc @jianlinliu 
or
/cc @jinyunma 

Could you PTAL at the installer-related changes ? Thank you.

**Comment by fxierh** [comment]
/cc @liangxia 
/cc @YuLi517 @heliubj18 

Could you PTAL ? Thank you.

**Comment by bryan-cox** [comment]
Hey @fxierh - I briefly looked this over and have a couple of questions/comments:
* Where ever we can, I recommend using multi-arch nightly images instead of amd64. Eventually we will have arm nodepool support before ARO HCP GA.
* I believe the ARO HCP team has some existing bicep templates they plan on using to create AKS mgmt clusters. Any reason we can't use those or coordinate with them on that topic? I would think we would want create AKS mgmt clusters in the same way they will be made in prod.

**Comment by fxierh** [comment]
Hi @bryan-cox

Thanks for the review. 

> Where ever we can, I recommend using multi-arch nightly images instead of amd64.

Yup, agreed. The catalog images annotated on the HC are multi-arch. While I filter by linux/amd64 when retrieving image info, ultimately it's the .listDigest that gets extracted and annotated on the HC. Filtering by a different platform and then extracting .listDigest results in the same hash.

> I believe the ARO HCP team has some existing bicep templates they plan on using to create AKS mgmt clusters.

I conducted a global search within this repo and found no existing infrastructure for AKS cluster creation. There might be something relevant elsewhere, but we would need to copy it here. Additionally, I checked with the installer team, and they are not aware of anything we could utilize in our case.

File: ci-operator/step-registry/hypershift/azure/create/hypershift-azure-create-chain.yaml [hunk 1]
```diff
@@ -54,10 +57,21 @@ chain:
         ENCRYPTION_KEY_ID="$(cat "${SHARED_DIR}/azure_active_key_url")"
       fi
       
-      /usr/bin/hypershift create cluster azure ${ICSP_COMMAND} \
+      EXTRA_ARGS=""
```
Comment (jianlinliu) [inline ci-operator/step-registry/hypershift/azure/create/hypershift-azure-create-chain.yaml:60]: Just to be curious, sounds like the content of the chian.yaml looks more like a ref step file, why here we are writing codes in a chain yaml file?

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+if [[ "$AKS_GENERATE_SSH_KEYS" == "true" ]]; then
+    AKE_CREATE_COMMAND+=(--generate-ssh-keys)
+fi
+
+if [[ "$AKS_ENABLE_FIPS_IMAGE" == "true" ]]; then
+    AKE_CREATE_COMMAND+=(--enable-fips-image)
+fi
+
+echo "Creating AKS cluster"
+eval "${AKE_CREATE_COMMAND[*]}"
```
Comment (jinyunma) [inline ci-operator/step-registry/aks/provision/aks-provision-commands.sh:35]: Just see provision step here to create aks cluster, but not find related deprovision step under step-registry/aks.

seems that az aks also support command to delete cluster.
```
$ az aks --help

Group
    az aks : Manage Azure Kubernetes Services.

Subgroups:
    command                        : See detail usage in 'az aks command invoke', 'az aks command
                                     result'.
    nodepool                       : Commands to manage node pools in Kubernetes kubernetes cluster.
    oidc-issuer                    : Oidc issuer related commands.

Commands:
    browse                         : Show the dashboard for a Kubernetes cluster in a web browser.
    check-acr                      : Validate an ACR is accessible from an AKS cluster.
    create                         : Create a new managed Kubernetes cluster.
    delete                         : Delete a managed Kubernetes cluster.
```
Do we need to add deprovision step to make sure that all resources created by aks are destroyed?

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/aks/hypershift/provision/cucushift-installer-rehearse-azure-aks-hypershift-provision-chain.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  - chain: cucushift-installer-rehearse-azure-aks-provision
+  - ref: cucushift-hypershift-extended-k8s-mgmt-apply-crds
+  - chain: cucushift-hypershift-extended-install-private
+  - ref: cucushift-hypershift-extended-k8s-mgmt-get-guest-annotations
+  - chain: hypershift-azure-create
+  - chain: cucushift-hypershift-extended-enable-qe-catalogsource
+  env:
+  - name: ENABLE_ICSP
+    default: "false"
+  - name: HYPERSHIFT_AZURE_LOCATION
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/aks/hypershift/provision/cucushift-installer-rehearse-azure-aks-hypershift-provision-chain.yaml:13]: Maybe an old topic, why we do not use `LEASED_RESOURCE` here but introduce a customized region here? I can not remember the reason.

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh (hunk 2 above)
Comment (jianlinliu) [reply ci-operator/step-registry/aks/provision/aks-provision-commands.sh:35]: Agree, it is better to introduce deprovision chain as a pair.

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/aks/hypershift/provision/cucushift-installer-rehearse-azure-aks-hypershift-provision-chain.yaml (hunk 3 above)
Comment (fxierh) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/aks/hypershift/provision/cucushift-installer-rehearse-azure-aks-hypershift-provision-chain.yaml:13]: Totally agree. I've already submitted a PR which default region to LEASED_RESOURCE for AWS by default. I also plan to do the same for Azure (in a separate PR). 

Thanks.

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.17-periodics-e2e-azure-conformance

File: ci-operator/step-registry/cucushift/hypershift-extended/health-check/cucushift-hypershift-extended-health-check-commands.sh [hunk 4]
```diff
@@ -129,7 +129,10 @@ fi
 echo "check mgmt cluster's HyperShift part"
 if test -s "${SHARED_DIR}/mgmt_kubeconfig" ; then
   export KUBECONFIG=${SHARED_DIR}/mgmt_kubeconfig
-  print_clusterversion
+  # Print clusterversion only if the management cluster is an OpenShift cluster
+  if oc get ns openshift; then
+      print_clusterversion
+  fi
```
Comment (LiangquanLi930) [inline ci-operator/step-registry/cucushift/hypershift-extended/health-check/cucushift-hypershift-extended-health-check-commands.sh:135]: According to the current rules, `mgmt_kubeconfig` must be an OpenShift cluster, so we do not need to add this check.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-nightly.yaml [hunk 5]
```diff
@@ -1364,6 +1364,17 @@ tests:
...
+  cron: 24 8 3,10,17,26 * *
+  steps:
+    cluster_profile: azure-qe
+    env:
+      HYPERSHIFT_HC_VERSION: "4.16"
+      TEST_FILTERS: ~ChkUpgrade&;~DisconnectedOnly&;~NonPreRelease&;~HyperShiftMGMT&;~MicroShiftOnly&;~NonHyperShiftHOST&;~Serial&;~Disruptive&
+      TEST_TIMEOUT: "30"
+    test:
+    - chain: openshift-e2e-test-hypershift-qe
+    workflow: cucushift-installer-rehearse-azure-aks-hypershift-guest
```
Comment (LiangquanLi930) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-nightly.yaml:1377]: It looks like we only have the CI for guest. Don't we need to add the CI for mgmt as well?

File: ci-operator/step-registry/cucushift/hypershift-extended/install-private/cucushift-hypershift-extended-install-private-commands.sh [hunk 6]
```diff
@@ -17,30 +14,79 @@ if [[ $HO_MULTI == "true" ]]; then
...
+    ;;
+*azure*)
+    if [[ -n "$HYPERSHIFT_EXTERNAL_DNS_DOMAIN" ]]; then
+        COMMAND+=(
+            --external-dns-credentials=/etc/hypershift-ext-dns-app-azure/credentials
+            --external-dns-provider=azure
+            --external-dns-domain-filter="$HYPERSHIFT_EXTERNAL_DNS_DOMAIN"
+        )
+    fi
+    cat /etc/hypershift-ext-dns-app-azure/credentials
```
Comment (heliubj18) [inline ci-operator/step-registry/cucushift/hypershift-extended/install-private/cucushift-hypershift-extended-install-private-commands.sh:82]: why cat the credentials here?

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-nightly.yaml (hunk 5 above)
Comment (heliubj18) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-nightly.yaml:1377]: I discussed it with @fxierh , the mgmt is not an OCP cluster, so we would not add mgmt ci this time.
Just avoid spending too much effort to adjust the OCP cases in the aks.

File: ci-operator/step-registry/cucushift/hypershift-extended/install-private/cucushift-hypershift-extended-install-private-commands.sh (hunk 6 above)
Comment (fxierh) [reply ci-operator/step-registry/cucushift/hypershift-extended/install-private/cucushift-hypershift-extended-install-private-commands.sh:82]: I forgot to remove this line. Thanks.

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh (hunk 2 above)
Comment (fxierh) [reply ci-operator/step-registry/aks/provision/aks-provision-commands.sh:35]: @jinyunma @jianlinliu 

Good idea. Updated.

File: ci-operator/step-registry/cucushift/hypershift-extended/health-check/cucushift-hypershift-extended-health-check-commands.sh (hunk 4 above)
Comment (LiangquanLi930) [reply ci-operator/step-registry/cucushift/hypershift-extended/health-check/cucushift-hypershift-extended-health-check-commands.sh:135]: Could you help to share the failed link?

**Comment by fxierh** [comment]
Another conflict ...

**Pull request merged**